osc/commandline.py

```diff
@@ -6406,21 +6406,27 @@ def parse_repoarchdescr(self, args, noinit = False, alternative_project = None,
                 # only persist our own repos
                 Repo.tofile(repolistfile, repositories)
 
-        no_repo = False
-        repo_names = sorted(set([r.name for r in repositories]))
-        if not arg_repository and repositories:
+        no_repo = not not arg_repository
+        if no_repo:
+            repo_arches = {}
+            for repo in repositories:
+                repo_arches.setdefault(repo.name, []).append(repo.arch)
+            repo_names = sorted(repo_arches.keys())
             # XXX: we should avoid hardcoding repository names
             # Use a default value from config, but just even if it's available
             # unless try standard, or openSUSE_Factory, or openSUSE_Tumbleweed
             no_repo = True
-            arg_repository = repositories[-1].name
-            for repository in (conf.config['build_repository'], 'standard', 'openSUSE_Factory', 'openSUSE_Tumbleweed'):
-                if repository in repo_names:
-                    arg_repository = repository
+            repo_order = [conf.config['build_repository'], 'standard', 'openSUSE_Factory', 'openSUSE_Tumbleweed']
+            repo_order.extend(reversed([r for r in repo_names]))
+            for repo_name in repo_order:
+                if arg_arch in repo_arches.get(repo_name, []):
+                    arg_repository = repo_name
                     no_repo = False
                     break
+            if no_repo:
+                raise oscerr.WrongArgs('%s is not a valid arch for repositories: %s' % (arg_arch, ', '.join(repo_order)))
 
-        if not arg_repository:
+        if no_repo:
             raise oscerr.WrongArgs('please specify a repository')
         if not noinit:
             if not arg_repository in repo_names:
```

Comment:
```
  File ".../osc/commandline.py", line 6432, in parse_repoarchdescr
    if not arg_repository in repo_names:
UnboundLocalError: local variable 'repo_names' referenced before assignment
```
After adding `repo_names = []` before `if no_repo: ` on line 6410, I got:
```
None is not a valid repository, use one of: 
```

-> This whole block must be moved before line 6410 instead:
```
        repo_arches = {}
        for repo in repositories:
            repo_arches.setdefault(repo.name, []).append(repo.arch)
        repo_names = sorted(repo_arches.keys())
```

And also  `no_repo = not not arg_repository` must be changed to `no_repo = not arg_repository` otherwise the provided repo is ignored.

Reply:
how do you get the errors described?

Reply:
`no_repo = False`  means that initial argument was `None`. It does't meant that detecte `arg_repository` is empty. I would leave it alone.